### DISCUSSION TITLE
XML Example

### DISCUSSION THREAD

**Discussion by yesoreyeram** [grafana/grafana-infinity-datasource#37] [Show and tell]
## XML Example

URL : https://gist.githubusercontent.com/yesoreyeram/655a362eed0f51be24e16d3f1127a31d/raw/aa58549a5cf9d06dae1204b5a09be5d651adc744/text.xml

As you can see in the below image, the above xml can be visualised in grafana table in few simple steps. First you need to specify url of the xml, then specify the root/rows value and then finally columns.

If your rows are directly under the root node, you can specify the rows as shown below. In my example I want to list all the CDs in the CATALOG. So my root node is `CATALOG.CD`. Then for columns, I can specify each column and its type as shown below

![image](https://user-images.githubusercontent.com/153843/108432478-58e5ee00-723c-11eb-8af9-052dde9c55aa.png)

## XML Example with time columns

With the same url above, we are going to parse the Year as time column.

![image](https://user-images.githubusercontent.com/153843/108433445-ea099480-723d-11eb-94b6-36c558149cee.png)

## XML Example with time column

URL : https://thingspeak.com/channels/38629/feed.xml 

in the below example, I want to visualise thingspeak traffic sensor data. As per the above xml, my feed data is two level down to the root node. So my root/rows selector is `channel.feeds[0].feed`. So I am trying to list all the feed data. 

![image](https://user-images.githubusercontent.com/153843/108431373-af522d00-723a-11eb-9954-ca0bdefe4ff1.png)

In the above example, you can see timestamp have very special selector `created-at[0]._`. because that node have a very specific property called `type="dateTime"` 

## XML with deep level of nodes

![image](https://user-images.githubusercontent.com/13445152/106956431-40a3a880-673f-11eb-84d1-297022a6e675.png)

In the above example, user used `xml.meter[0].profile[0].sample` as the root node

![image](https://user-images.githubusercontent.com/13445152/106956684-a132e580-673f-11eb-94a3-c977475caa30.png)

## XML - Simulating Timeseries

In the below example we don't have any time column. With just string column and one or more number columns, we can format the data as timeseries and then visuzalise in grafana's timeseries panels like guage, bar chart, stats panel etc.

![image](https://user-images.githubusercontent.com/153843/108434283-50db7d80-723f-11eb-858e-25a7fef18edb.png)

**Comment by PolliniG** [comment]
Hi how are you? very good your git. I need your help.
Can I do this with my infinity?
![image](https://user-images.githubusercontent.com/81996932/123651153-9f0e1800-d801-11eb-830f-dfef382cc819.png)

**Reply by yesoreyeram** [reply]
it seems you are using older version of infinity.

**Reply by PolliniG** [reply]
Perfect, thanks. I have this xml url, what would its root be?.
https://servicios1.afip.gov.ar/wsfe/service.asmx/FEDummy
I can't see it

**Reply by yesoreyeram** [reply]
since this is not a array data, you can simply define fields like this. 
![image](https://user-images.githubusercontent.com/153843/123671844-9e43a900-d836-11eb-9aa0-55cd83bc2b0d.png)
for the following xml

```xml
<DummyResponse xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns:xsd="http://www.w3.org/2001/XMLSchema" xmlns="https://xxxx">
    <appserver>OK</appserver>
    <dbserver>OK</dbserver>
     <authserver>OK</authserver>
</DummyResponse>
```

**Reply by PolliniG** [reply]
Sorry for being repetitive, but it does not work for me with the following parameters.

![image](https://user-images.githubusercontent.com/81996932/123672891-3b421a00-d816-11eb-8b64-c72f4050ff29.png)

**Reply by yesoreyeram** [reply]
As i said, you are using older version of the plugin. You may need to update. Also change the Viz type to table to confirm if you are reading data. ( Note:  Your data structure doesn't have any time stuff. So will not work with graphs. )

**Reply by PolliniG** [reply]
Hello friend, update the version of infinity and now it returns me this message
![image](https://user-images.githubusercontent.com/81996932/124009538-1edbe400-d9b4-11eb-9376-07ec05aa1949.png)

**Reply by yesoreyeram** [reply]
Sorry for the delayed response. Can you check the query inspector which gives you the corresponding curl command executed?. With the curl command you can manually verify that the plugin builds a correct url or not.

**Comment by akang-donai** [comment]
Hi, I have sample like in "XML with deep level of nodes", when I [tried](https://light.mizutech.id/data/camera/UIKKJ.20211118.212111.202018001747.xml), it doesn't work.

**Reply by yesoreyeram** [reply]
this one? (but the date field, I couldn't get as timestamp. also Object nodes differ on each item.. So doesn't recognized as array ). But just able to retrieve the data.

![image](https://user-images.githubusercontent.com/153843/142469693-8447b566-8797-4172-9801-9d29739b2b0a.png)

**Reply by akang-donai** [reply]
I See, do the timestamp mandatory?

**Reply by akang-donai** [reply]
One more thing, the URLs, can it be with wildcard? because the XML filename are different, generated by the IoT

**Reply by yesoreyeram** [reply]
timestamp is mandatory if you want to render data in certain panels like timeseries. for table kind of panels doen't.

URLs can't be wildcards. There is no way grafana know how to expand them into actual url.

**Comment by MKardin** [comment]
Hi,  I have a XML-file in a format with some <test-suite> </test-suite>
I view the first line in the table, but not the data for the second, third and fourth data-set. 
What can I do to see the 4 data-sets?

![image](https://user-images.githubusercontent.com/102597990/160612938-a2100ed3-2550-4775-bc1f-71fcfabaf03a.png)
[report.txt](https://github.com/yesoreyeram/grafana-infinity-datasource/files/8371464/report.txt)

**Reply by yesoreyeram** [reply]
I think you need to fix the root selector.  

BTW, I prefer UQL over this as UQL give more flexibility around parsing.Following UQL query worked for me. Hope this give you an idea.

```kusto
parse-xml
| scope "report-summary.test-suite"
| project "Time"=unixtime_milliseconds_todatetime("summary.@_timestamp"), "Host"="@_host", "Duration"=tonumber("summary.@_duration"), "Errors"=tonumber("summary.@_errors")
```

<img width="1777" alt="image" src="https://user-images.githubusercontent.com/153843/160618228-148911a3-e32a-4c52-ac0c-050e9a061630.png">

**Reply by MKardin** [reply]
Thank you for your idea to take UQL. It looks ok. I copied your statement and get an query error. I've tried to find informations about UQL-statement but I don't find something practical. Can you please have a look at my UQL query and find the difference to yours? Can you please give me a link for informations about UQL?

![image](https://user-images.githubusercontent.com/102597990/162777547-653acf47-f966-4bd3-bb61-bf4129849e8b.png)

**Reply by gdiana** [reply]
You must use
parse-xml
| scope "report-summary.test-suite"
| project "T....    

and not

parse-xml
scope "report-summary.test-suite"
project "T

You have not included the |

**Reply by MKardin** [reply]
thank you for the hint.
I copy and paste now all, but it makes query error.
![image](https://user-images.githubusercontent.com/102597990/163373604-976702e5-35f6-4fe5-9816-4d5e307279db.png)


As data I took the lines in report.xml and as UQL-Statement I took your input.
[report.txt](https://github.com/yesoreyeram/grafana-infinity-datasource/files/8488484/report.txt)

**Reply by gdiana** [reply]
You are probably using windows, where UQL has a known error. This is not the case on MAC OS.

This has been fixed in the latest infinity datasource update(0.8.6) so try it. 

Also when you update rather  download the ZIP file, extract, copy and restart Grafana.

Do not use the cmd to do this as it is still linked to the old buggy version.

You can try the new # comment feature and if it does not work then your are on the old version

**Reply by MKardin** [reply]
Yes, we are using windows. So we will try the latest infinity datasource update (0.8.6).

**Reply by gdiana** [reply]
👍

**Reply by MKardin** [reply]
Yesterday, we installed the latest datasource update (0.8.8). It appears the Query error message. 
I try the  new # comment feature in the uql-statement, but the query error message appears. 
In attach the Data,  UQL-Statement and a video with the comment feature.
[Explore_DataAndUQL.txt](https://github.com/yesoreyeram/grafana-infinity-datasource/files/8563676/Explore_DataAndUQL.t

https://user-images.githubusercontent.com/102597990/165313126-4fdb5db7-8d30-4c9c-86ed-28deceb632b1.mp4

xt)

**Reply by yesoreyeram** [reply]
It doesn't seems you are using 0.8.8 ( may be cache effect/ require restart of grafana ). In 0.8.8 some UQL keywords will be shown in purple and blue as shown below. Can you try [this](https://grafana-infinity-datasource.herokuapp.com/explore?orgId=1&left=%7B%22datasource%22:%22Infinity%22,%22queries%22:%5B%7B%22refId%22:%22A%22,%22type%22:%22uql%22,%22source%22:%22inline%22,%22format%22:%22table%22,%22url%22:%22https:%2F%2Fjsonplaceholder.typicode.com%2Fusers%22,%22url_options%22:%7B%22method%22:%22GET%22,%22data%22:%22%22%7D,%22root_selector%22:%22%22,%22columns%22:%5B%5D,%22filters%22:%5B%5D,%22global_query_id%22:%22%22,%22data%22:%22%3C%3Fxml%20version%3D%5C%221.0%5C%22%20encoding%3D%5C%22ISO-8859-1%5C%22%3F%3E%5Cn%3Creport-summary%20basedir%3D%5C%22%5C%22%20duration%3D%5C%2276498%5C%22%20end-date%3D%5C%222022-03-28%2012:08:56%5C%22%20executed-by%3D%5C%22adm-kardinam%5C%22%20execution-date%3D%5C%222022-03-28%2012:01:51%5C%22%20host%3D%5C%22IT-CRMDEVT-01%5C%22%20java-version%3D%5C%221.8.0_292-b10%5C%22%20os-version%3D%5C%22amd64-Windows%20Server%202016-10.0%5C%22%20qftest-version%3D%5C%225.3.4%5C%22%20realtime%3D%5C%22424188%5C%22%20result%3D%5C%220%5C%22%20runid%3D%5C%22220328120151,%20220328120220,%20220328120458,%20220328120821%5C%22%20sourcedir%3D%5C%22%5C%22%20time%3D%5C%2212:01:51.998%5C%22%3E%5Cn%20%20%3Csummary%20duration%3D%5C%2276498%5C%22%20errors%3D%5C%220%5C%22%20exceptedtests%3D%5C%220%5C%22%20exceptions%3D%5C%220%5C%22%20executedtests%3D%5C%220%5C%22%20expectederrors%3D%5C%220%5C%22%20expectedfailedtests%3D%5C%220%5C%22%20failedtests%3D%5C%220%5C%22%20flakyerrors%3D%5C%220%5C%22%20flakyexceptions%3D%5C%220%5C%22%20nonimpltests%3D%5C%220%5C%22%20passedtests%3D%5C%220%5C%22%20realtime%3D%5C%22424188%5C%22%20skippedtests%3D%5C%220%5C%22%20skippedtestsets%3D%5C%220%5C%22%20tests%3D%5C%220%5C%22%20time%3D%5C%2212:01:51.998%5C%22%20warnings%3D%5C%220%5C%22%2F%3E%5Cn%20%20%3Ctest-suite%20basedir%3D%5C%22%5C%22%20executed-by%3D%5C%22adm-kardinam%5C%22%20execution-date%3D%5C%222022-03-28%2012:01:51%5C%22%20host%3D%5C%22IT-CRMDEVT-01%5C%22%20java-version%3D%5C%221.8.0_292-b10%5C%22%20name%3D%5C%22E:%2FQF-Test%2F20220328Marcel.qft%5C%22%20os-version%3D%5C%22amd64-Windows%20Server%202016-10.0%5C%22%20pstate%3D%5C%220%5C%22%20qftest-build%3D%5C%2223457%5C%22%20qftest-version%3D%5C%225.3.4%5C%22%20relname%3D%5C%22...%2F20220328Marcel.qft%5C%22%20runid%3D%5C%22220328120151%5C%22%20sourcefile%3D%5C%22...%2F220328120151.qzp%5C%22%20state%3D%5C%220%5C%22%20type%3D%5C%22qftest%5C%22%20vstate%3D%5C%220%5C%22%20xfile%3D%5C%22220328120151_report.xml%5C%22%3E%5Cn%20%20%20%20%3Csummary%20duration%3D%5C%2272%5C%22%20errors%3D%5C%220%5C%22%20exceptedtests%3D%5C%220%5C%22%20exceptions%3D%5C%220%5C%22%20executedtests%3D%5C%220%5C%22%20expectederrors%3D%5C%220%5C%22%20expectedfailedtests%3D%5C%220%5C%22%20failedtests%3D%5C%220%5C%22%20nonimpltests%3D%5C%220%5C%22%20passedtests%3D%5C%220%5C%22%20realtime%3D%5C%2272%5C%22%20skippedtests%3D%5C%220%5C%22%20skippedtestsets%3D%5C%220%5C%22%20tests%3D%5C%220%5C%22%20time%3D%5C%222022-03-28%2012:01:51%5C%22%20timestamp%3D%5C%221648461711998%5C%22%20warnings%3D%5C%220%5C%22%2F%3E%5Cn%20%20%3C%2Ftest-suite%3E%5Cn%20%20%3Ctest-suite%20basedir%3D%5C%22%5C%22%20executed-by%3D%5C%22adm-kardinam%5C%22%20execution-date%3D%5C%222022-03-28%2012:02:20%5C%22%20host%3D%5C%22IT-CRMDEVT-01%5C%22%20java-version%3D%5C%221.8.0_292-b10%5C%22%20name%3D%5C%22E:%2FQF-Test%2F20220328Marcel.qft%5C%22%20os-version%3D%5C%22amd64-Windows%20Server%202016-10.0%5C%22%20pstate%3D%5C%220%5C%22%20qftest-build%3D%5C%2223457%5C%22%20qftest-version%3D%5C%225.3.4%5C%22%20relname%3D%5C%22...%2F20220328Marcel.qft%5C%22%20runid%3D%5C%22220328120220%5C%22%20sourcefile%3D%5C%22...%2F220328120220.qzp%5C%22%20state%3D%5C%220%5C%22%20type%3D%5C%22qftest%5C%22%20vstate%3D%5C%220%5C%22%20xfile%3D%5C%22220328120220_report.xml%5C%22%3E%5Cn%20%20%20%20%3Csummary%20duration%3D%5C%228402%5C%22%20errors%3D%5C%220%5C%22%20exceptedtests%3D%5C%220%5C%22%20exceptions%3D%5C%220%5C%22%20executedtests%3D%5C%220%5C%22%20expectederrors%3D%5C%220%5C%22%20expectedfailedtests%3D%5C%220%5C%22%20failedtests%3D%5C%220%5C%22%20nonimpltests%3D%5C%220%5C%22%20passedtests%3D%5C%220%5C%22%20realtime%3D%5C%228402%5C%22%20skippedtests%3D%5C%220%5C%22%20skippedtestsets%3D%5C%220%5C%22%20tests%3D%5C%220%5C%22%20time%3D%5C%222022-03-28%2012:02:20%5C%22%20timestamp%3D%5C%221648461740259%5C%22%20warnings%3D%5C%220%5C%22%2F%3E%5Cn%20%20%3C%2Ftest-suite%3E%5Cn%20%20%3Ctest-suite%20basedir%3D%5C%22%5C%22%20executed-by%3D%5C%22adm-kardinam%5C%22%20execution-date%3D%5C%222022-03-28%2012:04:58%5C%22%20host%3D%5C%22IT-CRMDEVT-01%5C%22%20java-version%3D%5C%221.8.0_292-b10%5C%22%20name%3D%5C%22E:%2FQF-Test%2F20220328Marcel.qft%5C%22%20os-version%3D%5C%22amd64-Windows%20Server%202016-10.0%5C%22%20pstate%3D%5C%220%5C%22%20qftest-build%3D%5C%2223457%5C%22%20qftest-version%3D%5C%225.3.4%5C%22%20relname%3D%5C%22...%2F20220328Marcel.qft%5C%22%20runid%3D%5C%22220328120458%5C%22%20sourcefile%3D%5C%22...%2F220328120458.qzp%5C%22%20state%3D%5C%220%5C%22%20type%3D%5C%22qftest%5C%22%20vstate%3D%5C%220%5C%22%20xfile%3D%5C%22220328120458_report.xml%5C%22%3E%5Cn%20%20%20%20%3Csummary%20duration%3D%5C%2233513%5C%22%20errors%3D%5C%220%5C%22%20exceptedtests%3D%5C%220%5C%22%20exceptions%3D%5C%220%5C%22%20executedtests%3D%5C%220%5C%22%20expectederrors%3D%5C%220%5C%22%20expectedfailedtests%3D%5C%220%5C%22%20failedtests%3D%5C%220%5C%22%20nonimpltests%3D%5C%220%5C%22%20passedtests%3D%5C%220%5C%22%20realtime%3D%5C%2233513%5C%22%20skippedtests%3D%5C%220%5C%22%20skippedtestsets%3D%5C%220%5C%22%20tests%3D%5C%220%5C%22%20time%3D%5C%222022-03-28%2012:04:58%5C%22%20timestamp%3D%5C%221648461898110%5C%22%20warnings%3D%5C%220%5C%22%2F%3E%5Cn%20%20%3C%2Ftest-suite%3E%5Cn%20%20%3Ctest-suite%20basedir%3D%5C%22%5C%22%20executed-by%3D%5C%22adm-kardinam%5C%22%20execution-date%3D%5C%222022-03-28%2012:08:21%5C%22%20host%3D%5C%22IT-CRMDEVT-01%5C%22%20java-version%3D%5C%221.8.0_292-b10%5C%22%20name%3D%5C%22E:%2FQF-Test%2F20220328Marcel.qft%5C%22%20os-version%3D%5C%22amd64-Windows%20Server%202016-10.0%5C%22%20pstate%3D%5C%220%5C%22%20qftest-build%3D%5C%2223457%5C%22%20qftest-version%3D%5C%225.3.4%5C%22%20relname%3D%5C%22...%2F20220328Marcel.qft%5C%22%20runid%3D%5C%22220328120821%5C%22%20sourcefile%3D%5C%22...%2F220328120821.qzp%5C%22%20state%3D%5C%220%5C%22%20type%3D%5C%22qftest%5C%22%20vstate%3D%5C%220%5C%22%20xfile%3D%5C%22220328120821_report.xml%5C%22%3E%5Cn%20%20%20%20%3Csummary%20duration%3D%5C%2234511%5C%22%20errors%3D%5C%220%5C%22%20exceptedtests%3D%5C%220%5C%22%20exceptions%3D%5C%220%5C%22%20executedtests%3D%5C%220%5C%22%20expectederrors%3D%5C%220%5C%22%20expectedfailedtests%3D%5C%220%5C%22%20failedtests%3D%5C%220%5C%22%20nonimpltests%3D%5C%220%5C%22%20passedtests%3D%5C%220%5C%22%20realtime%3D%5C%2234511%5C%22%20skippedtests%3D%5C%220%5C%22%20skippedtestsets%3D%5C%220%5C%22%20tests%3D%5C%220%5C%22%20time%3D%5C%222022-03-28%2012:08:21%5C%22%20timestamp%3D%5C%221648462101675%5C%22%20warnings%3D%5C%220%5C%22%2F%3E%5Cn%20%20%3C%2Ftest-suite%3E%5Cn%3C%2Freport-summary%3E%5Cn%5Cn%5Cn%5Cn%5Cn%5Cn%5Cn%5Cn%5Cn%5Cn%22,%22uql%22:%22parse-xml%5Cn%7C%20scope%20%5C%22report-summary.test-suite%5C%22%5Cn%7C%20project%20%5C%22Time%5C%22%3Dunixtime_milliseconds_todatetime(%5C%22summary.@_timestamp%5C%22),%20%5C%22Host%5C%22%3D%5C%22@_host%5C%22,%20%5C%22Duration%5C%22%3Dtonumber(%5C%22summary.@_duration%5C%22),%20%5C%22Errors%5C%22%3Dtonumber(%5C%22summary.@_errors%5C%22)%22%7D%5D,%22range%22:%7B%22from%22:%22now-1h%22,%22to%22:%22now%22%7D%7D)?

<img width="1481" alt="image" src="https://user-images.githubusercontent.com/153843/165315565-5032156f-e2a8-4de1-8099-db87b42106e2.png">

**Reply by yesoreyeram** [reply]
also if you commented out the first line `parse-xml`, plugin doesn't know that it is xml.  so keep that line uncommented.

**Reply by MKardin** [reply]
Thanks. Now, we are using 0.0.8 and it works.

**Comment by hardeepsingh3** [comment]
Is there a way to use the XML type to query against DNS bind statistics channel? I keep getting a no data error when using the URL that is assigned to that channel. I understand that pointing it to the UI is not the same as using inline raw data

**Reply by yesoreyeram** [reply]
There is no much difference between inline and URL.. when using XML you need to select fields correctly.. or change the parser to backend to see data.. then work from there..

**Reply by hardeepsingh3** [reply]
Thanks for your reply, I was not able to fetch the data from the URL. I have added two screenshots, the first one showing the grafana dashboard not displaying any data with the URL I am using and the second screenshot showing the data the URL shows when accessed on my local browser. 
Screenshot 1:
![image](https://user-images.githubusercontent.com/87157686/227340421-e61d4b79-c652-40ba-b3c9-716fc1174feb.png)
Screenshot 2:
![image](https://user-images.githubusercontent.com/87157686/227341011-3900e86b-87f4-4db7-8bcc-de8ada22565d.png)

Is the DNS statistics-channels not compatible with this plugin?

**Comment by abhishek-kgn** [comment]
Are there any documentation for UQL for xml? 
I am new to this,

**Comment by gdiana** [comment]
Yes, you may use the UQL command parse-xml which will convert XML to JSON.

UQL uses JSON and the parse command converts other formats to JSON


It also allows you to use JSONATA which is also very powerful

Kind regards




Gregory Diana


_______________________________________


[https://ci3.googleusercontent.com/mail-sig/AIorK4w8GeDRCVvd0gKIJQRTNPIzZwaNVVO5gun5JIFvzJNUo-FsSEeTHErnI-bK3YQEvlPZXv8fsk4]







[https://mail.google.com/mail/u/0/?ui=2&ik=621c1f6941&attid=0.1&permmsgid=msg-f%3A1757970687899992335&th=18659106c9d4010f&view=fimg&fur=ip&sz=s0-l75-ft&attbid=ANGjdJ-LnFVYeAxg5roAQ9Rw7ul7w7DZQj-SJkiRXqk8BijUlyXFfBZC78SRV7XDB_FF103bDC0FAzps8amLSS3etoZbIH7zOjLsEUq_KY7_MlnmLmiT9nTSm0Ne5EM&disp=emb]



www.expoenergy.co.za
[https://mailtrack.io/trace/link/e82ab423593d9a0df17d6d51e93c685c913e9449?url=http%3A%2F%2Fwww.expoenergy.co.za&userId=1419085&signature=c6f337a62d2728d9] 


22 Glen Grove
Glenwood
Durban 4001


Phone(h&w):031-2616761 (h)
Fax2Email: 086 474 3450
Cell: 0784515277
E-mail: ***@***.*** ***@***.***
E-mail: ***@***.*** ***@***.*** 
Skype: gregory.diana
Print: ***@***.*** ***@***.*** 



[data:image/gif;base64,R0lGODlhAQABAIAAAP///wAAACwAAAAAAQABAAACAkQBADs=3D]

**Reply by abhishek-kgn** [reply]
Thanks that helped.
But now I am stuck with below task;

I am have a column that has IPs from two subnets; and I need to replace that as shown below
eg.
11.123.234.xxx --->AB12
11.123.56.xxx --->C34

I tried using substr and replace, but it looks like that do not work together in replace when I did something "Final_col"=replace_string("substring("col1",1,5)",'src','replacer');
how can this be done using jsonata?

**Reply by gdiana** [reply]
To do this you can use the JSONATA EXERCISER or editor. You can use an online XML to JSON converter and then cut and paste JSON into the JSONATA editor and see what is required to do what you want to do.

**https://try.jsonata.org/**

Note in JSONATA editor you may use multiple lines but in infinity UQL it must be a single line

**Reply by gdiana** [reply]
You can use the JSONATA Execciser (https://try.jsonata.org/).

Cut and paste JSON in LH Window. Enter JSONATA commands in the top RH window and display the result in the bottom RH window.

It is a very powerful tool but requires a lot of learning.

Not in the Excerciser one may use multiple lines but in Infinity UQL it must all be on a single line.

**Reply by yesoreyeram** [reply]
You can also use the XML+BackendParser+MultilineJsonata as a root selector.

**Reply by gdiana** [reply]
That is great to know. I have not used it for some time now.
